stream-chat-android-state/src/main/java/io/getstream/chat/android/offline/plugin/logic/channel/internal/TypingEventPruner.kt

```diff
@@ -0,0 +1,226 @@
+/*
+ * Copyright (c) 2014-2022 Stream.io Inc. All rights reserved.
+ *
+ * Licensed under the Stream License;
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    https://github.com/GetStream/stream-chat-android/blob/main/LICENSE
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.getstream.chat.android.offline.plugin.logic.channel.internal
+
+import io.getstream.chat.android.client.events.TypingStartEvent
+import io.getstream.chat.android.client.models.ConnectionState
+import io.getstream.chat.android.client.models.TypingEvent
+import io.getstream.chat.android.client.setup.InitializationCoordinator
+import io.getstream.chat.android.core.internal.InternalStreamChatApi
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+
+/**
+ * A class used to keep track of typing events
+ * and remove stale ones that were not explicitly stopped by the sender
+ * due to technical difficulties. (e.g. process death, loss of Internet
+ * connection)
+ *
+ * @param channelId The id of the channel for which the typing events are being pruned.
+ * @property coroutineScope The scope used to launch jobs.
+ * @param delayTimeMs The period of time it takes before an individual stale
+ * typing event is removed.
+ * @param onUpdated Triggered when typing events have been updated.
+
+ */
+internal class TypingEventPruner(
+    private val channelId: String,
+    private val coroutineScope: CoroutineScope,
+    private val delayTimeMs: Long = DEFAULT_DELAY_TIME_MS,
+    initializationCoordinator: InitializationCoordinator = InitializationCoordinator.getOrCreate(),
```

Comment:
Why do you need this instance?
I can see you needed to create a new method to provide the current `ConnectionState`, but from the place where `TypingEventPruner` is created you have access to the `ClientState` that provides a property `connectionState` with the same value.
Can't we use it instead?
On the other hand, I can see it is used to stop previous launched job, but, Is it right?
Let me show you an example:
-. We receive a typing indicator from another user.
-. Our device switch from one network to another (There is an small disconnection)
-. Why did we clear typing indicators?

Reply:
I think we could use `ClientState`, and since `connectionState` is a `StateFlow` we could use `collectLatest` with a slight delay (100ms let's say) which would means that small interruptions to the connectivity should not clear the typing indicators.


Btw. eagerly clearing typing indicators isn't such a huge deal because in the current mode of operation, they are sent every three seconds for as long as someone is typing.